### ISSUE TITLE
Promise.cancel()

### ISSUE THREAD

**Issue by Artazor** [petkaantonov/bluebird#554]
Would it be consistent in 3.0 to add static `Promise.cancel()` to return a new promise in cancelled state? (like `Promise.resolve()` and `Promise.reject()`)

```JavaScript
Promise.cancel = function() {
      var p = Promise(function(){});
      p.cancel();
      return p;
}
```

**Comment by petkaantonov** [comment]
There is no cancellation state

**Comment by benjamingr** [comment]
To add, you can Promise.reject with a CancellationError

**Comment by petkaantonov** [comment]
@benjamingr cancellation isn't rejection with cancellation in 3.0

**Comment by Artazor** [comment]
@benjamingr it would be a rejection not a cancellation, wouldn't it? (in 3.0)

@petkaantonov - Can promise make a decision to cancel itself? (e.g. in `.then`) or the cancellation should be initiated only externally? 

Is the following case valid?
```JavaScript
...).then(function(...){
    if (cancellationCondition) {  

        // now there is no reason to continue (resolve/reject)
        // we want only perform a cleanup

        return Promise.cancel()   // as defined above
    } else {
        // Normal flow
    }
})
```

**Comment by bergus** [comment]
@petkaantonov 
> cancellation isn't rejection with cancellation in 3.0

But it should be for those handlers that are not in the cancelled chain. If a promise is cancelled, and then a new handler is added, the new handler should see a rejection with `CancellationError`.

@Artazor 
> Can promise make a decision to cancel itself?

No. Cancellation is initiated from outside. A promise can decide to cancel a promise on which it does depend itself (e.g. `timeout` would do that), but to its subscribers only a rejection can be signalled.

An example (for both of my points):
```
class TimeoutError extends CancellationError { }
Promise.prototype.timeout = function(ms) {
    setTimeout(() =>
        this.cancel(new TimeoutError);
    );
    return this; // this.then() to return a new instance, but without cancellation
                 // supporting multiple consumers it doesn't matter anyway
};
```
then `fetch(…).timeout(5000)` would be rejected after 5s.
And similarly `delay(6000).then(function thatIsNeverExecuted(){}).timeout(5000).catch(function thatIsExecuted(){})`

**Comment by spion** [comment]
@bergus 

> If a promise is cancelled, and then a new handler is added, the new handler should see a rejection with CancellationError.

That seems pretty bad to me: rejection handlers may or may not be called for cancelled promise depending on the timing

**Comment by bergus** [comment]
@spion No, they would *always* get called. Except a) the promise fulfills before it is cancelled or b) they are in a chain that is itself cancelled before the rejection.

This is pretty consistent: Your handlers are always called on resolution, if you have not cancelled them yourself. *No* outside factors at all.
What would your expectations be in the following case:

    var p = fetch(…); // takes a second
    p.timeout(500).…; // implemented as above, for simplicity (cancelling p after the timeout)
    p.then(a, b).…;

Would `a` get called after one second? Would `b` get called after half a second? Would neither ever be called?

**Comment by spion** [comment]
@bergus Hmm I thought that in 3.0 cancellation will not result with rejections; only finally handlers will be called, but not fulfillment or rejection handlers. Or was that idea abandoned?

**Comment by Artazor** [comment]
@spion the idea was not abandoned! It's probably my fault that @bergus thinks that Cancellation is Rejection in 3.0 (due to my comments in discussion #415). Really there was a moment when @petkaantonov wanted to postpone `forget` semantics until 4.0. He even has changed the label from 3.0 to 4.0 (at Feb-7), but then he reverted it back to 3.0 (at Mar-28).
So we have a new cool implementation that is almost ready: @petkaantonov is committing right now

**Comment by bergus** [comment]
@spion, @Artazor no, please don't understand me wrong. I do want forget semantics, I really like them, I do endorse them, I've even written [a draft proposal](https://gist.github.com/bergus/c160b1c0c9e755108d1b) for how they should work. Let's take
```
var p = fetch().then(onful, onrej).finally(onfin);
p.cancel();
```
then neither `onful` nor `onrej` should be called, `onfin` would, and the fetch would be aborted.

But what is missing from this simple description: How would branching be handled? That's what I've also commented on [here](415#issuecomment-81978123). Proceeding with the example where `p` is already cancelled, what happens when we do
```
p.then(a, b);
```
*after the fact*? Should those handlers simply be ignored? Is it a truly forever-pending promise now? Imo, `b` should be called with a `CancellationError` now. What do you think?

**Comment by petkaantonov** [comment]
Cancellation won't cause a rejection regardless at what point you attach a handler

**Comment by bergus** [comment]
So if multiple consumers are interested in a promise, and one wants to cancel his subscription, the others don't even get notified that they will never resolve? This doesn't seem to be a good idea. And sounds to me much like a new and unobservable "cancelled" state that propagates down a chain.

**Comment by spion** [comment]
@bergus you do understand that what you propose intintroduces the timing issue right? Launch two operations in parallel then cancel one of them then from the other wait for the cancelled one: what happens? The answer is: it depends on how long they both took

**Comment by bergus** [comment]
@spion: Sorry I'll have to translate that into code to make sure we're talking of the same thing.
```
var a = operationOne();
var b = operationTwo();
setTimeout(() => a.cancel(), 1000) // not cancelled immediately as that would be trivial :-)
var c = b.then(() => a)
```
My proposal would ensure that `c` is (assuming `b` fulfills) either fulfilled with `a`s result (when `a` took less than a second to fulfill), or rejected with a `CancellationError` (when `a` was cancelled before it could resolve). Regardless how long `b` took.
Unless someone did `c.cancel()` of course :-)

You are right, cancellation is a complex thing and it will be possible to construct race conditions when promises mutually cancel each other and similar things.
My point is that you will want to know when *someone else* cancelled a promise you are depending on. No debugging nightmare "why are my handlers never called?" - "oh, somehow a promise was cancelled mid-chain and I forgot `.finally` handlers", but an actual error "someone messed with your promises" that can be caught. Of course it's not an error when you triggered the cancellation yourself (i.e. `.cancel()` the head of your own chain), then the forget semantics apply.

**Comment by bergus** [comment]
@spion Another code example of an actual race where cancellation depends on how long an action took:
```
var a = Promise.delay(Math.random()*2000).then(()=> b.cancel()),
    b = Promise.delay(Math.random()*2000).then(()=> a.cancel());
a.then(()=>console.log("a wins"), ()=>console.log("a loses"));
b.then(()=>console.log("b wins"), ()=>console.log("b loses"));
```
Notice how the forget/ignore behaviour of cancellation ensures that only one of the `b.cancel()`/`a.cancel()` callbacks is ever executed. So exactly one will win, and the other will lose (instead of forever awaiting notice of the outcome :-P).

Oh, and I did delibaretely not use
```
var a = Promise.delay(Math.random()*2000),
    b = Promise.delay(Math.random()*2000);
a.then(()=> b.cancel()).then(()=>console.log("a wins"), ()=>console.log("a loses"));
b.then(()=> a.cancel()).then(()=>console.log("b wins"), ()=>console.log("b loses"));
```
where it is possible (albeit unlikely) that both delays resolve at the "same time" without executing their handlers in between, so that both a and b might win (or both might lose, when `.cancel()` throws on settled promises).

**Comment by petkaantonov** [comment]
@bergus what's wrong with requiring a `.cancel()` from all branched consumers? 

E.g.

```js
var a = request();
var branch1 = a.then()...;
var branch2 = a.then()...

branch1.cancel() // Propagates to a which increments its cancellation count by 1 but it requires 2
branch2.cancel() // Now the count is 2 out of 2 so the request can be e.g. aborted if desired
```

**Comment by Artazor** [comment]
> @bergus what's wrong with requiring a .cancel() from all branched consumers? 

@petkaantonov, at first look it seems like a great feature, since branching/joining (via `join`/`all`/`map`/etc) graph is always a DAG, so the reference counting here could matter. But unfortunately it will introduce the timing issue:

```JavaScript
var a = request();
var branch1 = a.then()...;
var branch2 = a.then()...

// and now start the race between two concurrent promises `p` and `q` 
p.then(function(){
    var branch3 = a.then()...  // will this branch depends on the race winner...
});

q.then(function(){
    branch1.cancel();
    branch2.cancel(); // are there any other branches attached to `a`?
});
```
I think that it is what @spion was afraid of. I'm afraid too... But anyway it looks more useful than harming. \[It is not implemented for now, but must be decided before the launch 'cause it will be a breaking change \]

**Comment by petkaantonov** [comment]
The thing is that branching like that is incredibly rare, cancellation is not often needed either, and combining both is just absurd to begin with. So this is why it doesn't really matter.

**Comment by bergus** [comment]
@petkaantonov 
> what's wrong with requiring a `.cancel()` from all branched consumers?

Nothing! That would be the optimal implementation imo :-) However, it increases complexity and adds overhead which might slow things down. My token idea would have implemented just that (on cancellation attempt, you'd count how many non-cancelled tokens are still registered), which gives you rather fine-grained control about how cancellation happens.
It just seemed [you didn't like this](415#issuecomment-81942558), and I'm fine with that - I don't really know if we need to provide a way for consumers to prevent others from cancelling. It's just that an *unexpected* cancellation should be an error imo.

And even with requiring cancellation attempts from all consumers, there's still the problem with consumers who are *late to the party*. For them, the already-happened cancellation is definitely unexpected. @Artazor has a nice example for this:
```
var a = request();
var branch1 = a.then(brancher1);

// and now start the race between two concurrent promises `p` and `q` 
p.then(function(){
    var branch2 = a.then(brancher2);
});

q.then(function(){
    branch1.cancel(); // are there any other branches attached to `a`?
});
```
In my proposal, the behaviour is as follows:
* `branch1` will be cancelled when `q` fulfills, that means `brancher1` will never be executed if that happens before `a` resolves
* `a` will be cancelled if `q` fulfills before `p` (yes, that's the race condition)
* `branch2` will either resolve with the result from `brancher2` (when `p` was faster than `q`), or be rejected with a CancellationError (otherwise).

Btw, one could easily avoid the race by doing
```
var aSafe = a.then(); // branch immediately - "register" to prevent `a.cancel()` from
                      // suceeding without also `aSafe.cancel()`
p.then(function(){
    var branch2 = aSafe.then(brancher2);
});
```

**Comment by petkaantonov** [comment]
Also any code that would even be affected by the @spion's race condition would also get false positives from unhandled rejection since it implies asynchronously attaching handlers

**Comment by spion** [comment]
@petkaantonov but you can mitigate that by immediately adding error handlers to that promise as well. Asynchronously adding handlers can be useful and PUR detection doesn't necessarily exclude it (as long as you handle the errors upfront you're good to go)

**Comment by petkaantonov** [comment]
@spion oh rigth. Anyway, just finished cancellation in 3.0 :) https://github.com/petkaantonov/bluebird/commit/204fb94dbbca38e8e42b5ec1245ca36f2584f889

**Comment by spion** [comment]
@bergus you're right wrt debugging stuck promises, which is why I think that forget semantics is not a good idea at all, but oh well. I guess we'll see (well, I won't because I don't plan to use cancellation :)

**Comment by Artazor** [comment]
> So if multiple consumers are interested in a promise, and one wants to cancel his subscription, the others don't even get notified that they will never resolve? This doesn't seem to be a good idea. And sounds to me much like a new and unobservable "cancelled" state that propagates down a chain.

@bergus, in fact, this is observable with `onCancel+finally`. Chaining from cancelled promises with `.then(...).onCancel(...).finally(...)`  still works: only `onCancel` and `finally` are called. This means that the there is the "cancellation state", and I'm wondering why @petkaantonov insists that
> There is no cancellation state

However I see the breaking change: a48d63ff9c112c1d607dcd3acd86f170ed0a826c (even one test case has been removed)
@petkaantonov can you elaborate a48d63ff9c112c1d607dcd3acd86f170ed0a826c a bit? What's wrong with cancellation of chained steps? 

```JavaScript
function request(arg) {
     var handler;
     return new Promise(function(resolve) {
         handler = setTimeout(function() {
             resolve({body: arg});
         }, 100);
     }).onCancel(function(){
         clearTimeout(handler);
     }).finally(function(){
         handler = null;
     });
}

var A = "A";
var p = request(A).then(function(response) { // request to A takes 100ms
    var B = response.body.replace("A","B");                  
    return request(B); // request to B takes another 100 ms
}).then(function(response) {
    console.log("Fulfilled:", response);
}).onCancel(function(){
    console.log("Cancelled");
});

setTimeout(function(){  
    // cancel the chain after 150 ms
    // when request(A) is already fulfilled but request(B) is still pending
    p.cancel();
}, 150)  
```

I'd expected to see "Cancelled", (and request(B) should be aborted and cleaned up).
After your refactoring a48d63ff9c112c1d607dcd3acd86f170ed0a826c, I was afraid that my expectations could become wrong (and it would be a disaster for me). But looks that it works as expected. Did I miss something?

**Comment by Artazor** [comment]
Even more surprising is that now
```JavaScript
var a = request("A"); // from the previous request
var p = a.then(function(res) {
    console.log("fulfilled:", res);
    p.cancel(); // break the chain early
}).then(function() {
    console.log("Called")
});
```
produces 
```
fulfilled: {body: "A"}
```
so the chain is broken.

But the same code with `Promise.resolve({body: "A"})`
```JavaScript
var a = Promise.resolve({body: "A"});
var p = a.then(function(res) {
    console.log("fulfilled:", res);
    p.cancel(); // break
}).then(function() {
    console.log("Called")
});
```
produces:

```
fulfilled: {body: "A"}
Called
```
So the chain **is not broken**

O_o?  (bye-bye mocked requests in unit/integration tests?)
Worth it to be registered as a new issue?

**Comment by petkaantonov** [comment]
The cancel in second case doesn't do anything because the promise is already fulfilled, it was a bug that was fixed.

**Comment by Artazor** [comment]
Which promise are we speaking about?
in both cases when we are inside 
```JavaScript
var p = a.then(function(res){
   // here `a` is already fulfilled, and `p` is pending.
})` 
```

**Comment by petkaantonov** [comment]
`p` is pending but it's not pending on any asynchronous action so cancelling doesn't make sense, what would you cancel?

**Comment by Artazor** [comment]
I don't understand. In both cases `p.cancel` is performed when there are no any asynchronous action pending (in both cases corresponding promises are already fulfilled), so in both cases there is no actions to cancel above the cancellation point, but there are following chain, that should be either cancelled or not consistently in both cases (by design you've chosen)

**Comment by petkaantonov** [comment]
Ah I see, I guess I need to revert the change

**Comment by bergus** [comment]
@petkaantonov we don't want to stop asynchronous actions only. We want to cancel callbacks as well, preventing them from being executed. Even if an async action is uncancellable, we still can choose not the execute handlers on it.

**Comment by petkaantonov** [comment]
@bergus Yes I get it now, I am reverting the patch

**Comment by petkaantonov** [comment]
@Artazor please try again with latest 3.0 branch :)

**Comment by Artazor** [comment]
Now all is ok. Should I perform any "rebase" magic to the repo to straighten the history, or there is no need for #563 test anymore?

**Comment by petkaantonov** [comment]
I added back the p.break() test so no need for the new test :)
